q2_longitudinal/_utilities.py

```diff
@@ -400,40 +391,84 @@ def _regplot_subplots_from_dataframe(state_column, metric, metadata,
-        plot_control_limits=plot_control_limits, ax=axes[0],
-        xtick_interval=xtick_interval)
-    c.set_title('Group volatility comparison plot')
-    c = _set_xticks(c, metadata, state_column, states, xtick_interval)
-
     # plot individual groups' control charts
     colors = cycle(sns.color_palette(palette, n_colors=len(groups)))
-    for num in range(len(groups)):
-        group = groups[num]
-        group_md = metadata[metadata[group_column] == group]
+    cmap = {}
+    num = 1
+    for group, group_md in metadata.groupby(group_column):
+        color = next(colors)
+        cmap[group] = color
         c, gm, gs = _control_chart(
             state_column, metric, group_md, None, ci=ci, legend=False,
-            color=next(colors), plot_control_limits=True, ax=axes[num + 1],
+            color=color, plot_control_limits=plot_control_limits, ax=axes[num],
             palette=None, xtick_interval=xtick_interval)
         c.set_title('{0}: {1}'.format(group_column, group))
+        if spaghetti != 'no':
+            # plot group's sphaghetti on main plot and current subplot
+            for ax in [0, num]:
+                c = _make_spaghetti(
+                    group_md, state_column, metric, individual_id_column,
+                    states, ax=axes[ax], color=color, alpha=0.3,
+                    spaghetti=spaghetti)
         c = _set_xticks(c, group_md, state_column, states, xtick_interval)
+        axes[num].set_yscale(yscale)
+        num += 1
+
+    # plot all groups together, compare variances
+    c, global_mean, global_std = _control_chart(
+        state_column, metric, metadata, group_column, ci=ci, palette=cmap,
+        plot_control_limits=plot_control_limits, ax=axes[0],
+        xtick_interval=xtick_interval)
+    c.set_title('Group volatility comparison plot')
+    c = _set_xticks(c, metadata, state_column, states, xtick_interval)
+
+    c.set_yscale(yscale)
+
     plt.tight_layout()
 
     return chart, global_mean, global_std
 
 
+def _make_spaghetti(metadata, state_column, metric, individual_id_column,
+                    states, ax, color=None, alpha=1.0, spaghetti='no'):
+    for ind, ind_data in metadata.groupby(individual_id_column):
+        # optionally plot mean of replicates
+        if spaghetti == 'mean':
+            ind_data = ind_data.groupby(state_column).mean()
+            # add state_column back into dataframe since we use it for plotting
+            ind_data[state_column] = ind_data.index
+        # Adjust xticks on so that it follows a pseudo-categorical scale
+        # (e.g., 0, 1, 7, 200 would be plotted at even intervals on x axis)
+        # so that spaghetti aligns with seaborn pointplot x axis.
+        altered_states = _adjust_xticks(ind_data, state_column, states)
+
+        ax.plot(altered_states, ind_data[metric], alpha=alpha, c=color,
+                label='_nolegend_')
+    return ax
+
+
+def _adjust_xticks(ind_data, state_column, states):
```

Comment:
This name makes me *think* this method mutates something. I don't have a good answer for fixing that, but it's kind of like `_make_*` which definitely mutates.

Would `_get_xticks` work better? I would assume I could call a helper like that multiple times which implies it doesn't mutate anything.

Reply:
👍